common/Reducers.jsm

```diff
@@ -457,6 +457,182 @@ function Pocket(prevState = INITIAL_STATE.Pocket, action) {
   }
 }
 
+// Update spocs data inside discovery stream by handling action data
+// received on click of link menu options
+function DiscoveryStreamSpocs(state = {}, action) {
+  switch (action.type) {
+    case at.PLACES_LINK_BLOCKED:
+      return {
+        ...state.spocs,
+        data: state.spocs.data.spocs ? {
+          spocs: state.spocs.data.spocs.filter(s => s.url !== action.data.url),
+        } : {},
+      };
+    case at.PLACES_SAVED_TO_POCKET:
+      return {
+        ...state.spocs,
+        data: state.spocs.data.spocs ? {
+          spocs: state.spocs.data.spocs.map(item => {
+            if (item.url === action.data.url) {
+              return Object.assign({}, item, {
+                open_url: action.data.open_url,
+                pocket_id: action.data.pocket_id,
+              });
+            }
+            return item;
+          }),
+        } : {},
+      };
+    case at.DELETE_FROM_POCKET:
+    case at.ARCHIVE_FROM_POCKET:
+      return {
+        ...state.spocs,
+        data: state.spocs.data.spocs ? {
+          spocs: state.spocs.data.spocs.filter(s => s.pocket_id !== action.data.pocket_id),
+        } : {},
+      };
+    case at.PLACES_BOOKMARK_ADDED:
+      return {
+        ...state.spocs,
+        data: state.spocs.data.spocs ? {
+          spocs: state.spocs.data.spocs.map(item => {
+            if (item.url === action.data.url) {
+              const newSite = Object.assign({}, item);
+              delete newSite.bookmarkGuid;
+              delete newSite.bookmarkTitle;
+              delete newSite.bookmarkDateCreated;
+              return newSite;
+            }
+            return item;
+          }),
+        } : {},
+      };
+    case at.PLACES_BOOKMARK_REMOVED:
+      return {
+        ...state.spocs,
+        data: state.spocs.data.spocs ? {
+          spocs: state.spocs.data.spocs.map(item => {
+            if (item.url === action.data.url) {
+              const {bookmarkGuid, bookmarkTitle, dateAdded} = action.data;
+              return Object.assign({}, item, {
+                bookmarkGuid,
+                bookmarkTitle,
+                bookmarkDateCreated: dateAdded,
+              });
+            }
+            return item;
+          }),
+        } : {},
+      };
+    default:
+      return state;
+  }
+}
+
+// Update recommendations data inside discovery stream feeds by handling
+// action data received on click of link menu options
+function DiscoveryStreamFeeds(state = {}, action) {
+  switch (action.type) {
+    case at.PLACES_LINK_BLOCKED:
+      return {
+        ...state.feeds,
+        data: Object.keys(state.feeds.data).reduce((accumulator, feed_url) => {
+          accumulator[feed_url] = {
+            data: {
+              ...state.feeds.data[feed_url].data,
+              recommendations: state.feeds.data[feed_url].data.recommendations.filter(r => r.url !== action.data.url),
```

Comment:
I didn't look that closely at all the others, but it seems like a higher order function that takes a function returning new recommendations might clean this up:
```js
case BLOCKED:
  return HOF(recs => recs.filter(r => …));
case SAVED:
  return HOF(recs => recs.map(item => …));
```

Reply:
Here's what I was thinking from our chat earlier (totally untested! mostly just copy pasted stuff 😉 ):
```js
function DiscoveryStream(prevState, action) {
  const updateLinks = onLinks => {
      // Return if action data is empty, or spocs or feeds data is not loaded
      return !action.data || !prevState.spocs.loaded || !prevState.feeds.loaded ? prevState : {
        ...prevState,
        spocs: {
          ...prevState.spocs,
          data: prevState.spocs.data.spocs ? {spocs: onLinks(prevState.spocs.data.spocs)} : {},
        },
        feeds: {
          ...prevState.feeds,
          data: Object.keys(prevState.feeds.data).reduce((accumulator, feed_url) => {
            accumulator[feed_url] = {
              data: {
                ...prevState.feeds.data[feed_url].data,
                recommendations: onLinks(prevState.feeds.data[feed_url].data.recommendations),
              },
            };
            return accumulator;
          }, {});
        },
      };
  };

  switch(action.type) {
    case at.PLACES_LINK_BLOCKED:
      return updateLinks(recs => recs.filter(r => r.url !== action.data.url));
  }
}
```
This is assuming the same onLinks callback can be used for feed links and spocs.